chat_client/src/main.py

```diff
@@ -0,0 +1,234 @@
+
+    ...
+
+    Methods
+    -------
+    connect():
+        Ask the user for login name and creates chat service stub,
+        must be called before run().
+    change_login_name():
+        Change or set the user name
+    run():
+        Run chat client
+    disconnect():
+        Close channel, receiver thread and any other stuff opened.
+        Must be called at the end of object existance.
+    """
+
+    def __init__(self, host: str, port: int) -> None:
+        """Constructs all the necessary attributes for the client object.
+
+        Parameters
+        ----------
+            host : str
+                Host address(Address of the server)
+            port : int
+                Port number
+        """
+        self._is_connected = False
+
+        self._channel = None
+        self._stub = None
+        self._receiver = None
+
+        self._username = ""
+        self._connection_addr = f"{host}:{port}"
+        logging.debug("Chat client object created")
+
+    def connect(self) -> None:
+        """Ask for username and connect."""
+        if self._is_connected:
+            return
+        try:
+            self._channel = grpc.insecure_channel(self._connection_addr)
+            self._stub = chat_pb2_grpc.ChatServiceStub(self._channel)
+            self._handle_register()
+            self._login()
+        except ConnectionRefusedError as e:
+            logging.error("Cannot connect [%s]", e)
+        # except Exception as e:
+        #     logging.error("Cannot connect [%s]", e)
+        else:
+            self._is_connected = True
+            logging.info("Chat client connected")
+        finally:
+            self.disconnect()
+            
+    def _login(self) -> None:
+        username =self._get_username()
+        for _ in range(3):
+            req = chat_pb2.LoginUserRequest(login=username, password=getpass())
+            try:
+                self._stub.LoginUser(request=req)
+            except grpc.RpcError as rpc_error:
+                if rpc_error.code() == grpc.StatusCode.UNAUTHENTICATED:
+                    logging.info("Wrong creds, try again...", username)
+                else:
+                    raise rpc_error
+            else:
+                self._username = username
+                return
+        raise ConnectionRefusedError("Login failed")
+
+    def _get_username(self) -> str:
+        username = input("login: ").strip()
+        while not username:
+            username = input("login: ").strip()
+            logging.info("Login cannot be blank")
+        return username
+
+    def run(self):
+        """Run chat client"""
+        if not self._is_connected:
+            logging.error(
+                "chat client is disconnected. (You have to call connect() before run())"
+            )
+            return
+
+        self._open_stream_receiver()
+        self._start_chat()
+        self._close_stream_receiver()
+        
+    def _handle_register(self) -> None:
```

Comment:
Tired of doctrsings?)))
There is and below you missed them, you may not add them for now, anyway that's only PoC)

Reply:
Done